### PR TITLE
Add a proposal for better conversion from collection expression

### PR DESCRIPTION
**Pull request by 333fred**

### CODE REVIEW

File: proposals/collection-expression-better-conversion.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+If `E₁` and `E₂` have an identity conversion, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:
+- For every `ELᵢ`, `CE₁ᵢ` is at least as good as `CE₂ᵢ`, and
+- There is at least one i where `CE₁ᵢ` is better than `CE₂ᵢ`
+  
+Otherwise, neither set of element conversions is better than the other, and they are also not as good as each other.  
+Conversion comparisons are made using better conversion from expression if `ELᵢ` is not a spread expression. If `ELᵢ` is a spread expression, we use better conversion from the element type of the spread collection to `E₁` or `E₂`, respectively.
+
+`T₁` is a ***better collection conversion from expression*** than `T₂` if:
+- `T'₁` is better than `T'₂`, and the element conversions to `E₁` are at least as good as the element conversions to `E₂`, or
+- `T'₁` is at least as good as `T'₂`, and the element conversions to `E₁` are better than the element conversions to `E₂`.
```
Comment (cston) [inline proposals/collection-expression-better-conversion.md:53]:
```suggestion
- `T'₁` is the same as `T'₂`, and the element conversions to `E₁` are at least as good as than the element conversions to `E₂`, and at least one of the element conversions to `E₁` is better than the element conversion to `E₂`.
```

File: proposals/collection-expression-better-conversion.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    - If `T₁` is a generic type, `T₁` with all type parameters unsubstituted
+    - If `T₁` is an array type, an unsubstituted array type
+    - Otherwise, `T₁`
+- `T'₂` is the same as `T'₁`, except for using `T₂` instead of `T₁`
+- `E₁` is the element type of `T₁`
+- `E₂` is the element type of `T₂`
+- `CE₁ᵢ` are the series of conversions from `ELᵢ` to `E₁`
+- `CE₂ᵢ` are the series of conversions from `ELᵢ` to `E₂`
+
+`T'₁` is at least as good as `T'₂` if there is an identity conversion between them.
```
Comment (cston) [inline proposals/collection-expression-better-conversion.md:36]: It sounds like `T'₁` is at least as good as `T'₂` implies that `T'₂` is at least as good as `T'₁`. Perhaps we should use "is the same as" or similar term instead.
```suggestion
`T'₁` is the same as `T'₂` if there is an identity conversion between them.  
```

File: proposals/collection-expression-better-conversion.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+- `T'₂` is the same as `T'₁`, except for using `T₂` instead of `T₁`
+- `E₁` is the element type of `T₁`
+- `E₂` is the element type of `T₂`
+- `CE₁ᵢ` are the series of conversions from `ELᵢ` to `E₁`
+- `CE₂ᵢ` are the series of conversions from `ELᵢ` to `E₂`
+
+`T'₁` is at least as good as `T'₂` if there is an identity conversion between them.  
+`T'₁` is better than `T'₂` if:
+- `T'₁` is ReadOnlySpan<T>, and `T'₂` is not ReadOnlySpan<T>
+- `T'₂` is Span<T>, and `T'₂` is an *array_or_array_interface*
```
Comment (jnm2) [inline proposals/collection-expression-better-conversion.md:39]: This feels good to me.

File: proposals/collection-expression-better-conversion.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+- `T'₂` is implicitly convertible to `T'₁`, and `T'₁` is not implicitly convertible to `T'₂`.
+  
+Otherwise, `T'₁` neither as good as nor better than `T₂`.
```
Comment (jnm2) [inline proposals/collection-expression-better-conversion.md:42]:
```suggestion
Otherwise, `T'₁` neither as good as nor better than `T'₂`.
```

File: proposals/collection-expression-better-conversion.md (hunk 1 above)
Comment (jnm2) [reply proposals/collection-expression-better-conversion.md:53]: Not holding an opinion about which way, but I did notice there was something satisfying about the symmetry of "A is better and B is at least as good, or A is at least as good and B is better."

**Comment by jnm2** [comment]
The table is very hard to read on GitHub because only a fraction of the width is visible at a time.

File: proposals/collection-expression-better-conversion.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+### Scenarios:
+
+In plain English, the collection types themselves must be either the same, or unambiguously better (ie, `List<T>` and `List<T>` are the same, `List<T>` is unambiguously better than `IEnumerable<T>`, and `List<T>` and `HashSet<T>` cannot be compared), and
+the element conversions for the better collection type must also be the same or better (ie, we can't decide between `ReadOnlySpan<object>` and `Span<string>` for `[""]`, the user needs to make that decision). More examples of this are:
+
+
+| `T₁` | `T₂` | `E` | `C₁` Conversions | `C₂` Conversions | `T'₁` vs `T'₂` | `CE₁ᵢ` vs `CE₂ᵢ` | Outcome |
+|--------|--------|------------|----------------|----------------|---------------------|---------------------|---------|
+| `List<int>` | `List<byte>` | `[1, 2, 3]` | `[Identity, Identity, Identity]` | `[Implicit Constant, Implicit Constant, Implicit Constant]` | `T'₁` is as good as `T'₂` | `CE₁ᵢ` is better | `List<int>` is picked |
+| `List<int>` | `List<byte>` | `[(int)1, (byte)2]` | `[Identity, Implicit Numeric]` | Not applicable | `T'₁` is as good as `T'₂` | `T₂` is not applicable | `List<int>` is picked |
```
Comment (jnm2) [inline proposals/collection-expression-better-conversion.md:64]: It looks like `-` is used instead of `Not applicable` in the other rows. For clarity, what do you think of using `N/A` everywhere?

**Review by jnm2 — APPROVED** [review]

File: proposals/collection-expression-better-conversion.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+> **`E` is a *collection expression*, and `C₁` is a ***better collection conversion from expression***, or**
+> * **`E` is not a *collection expression* and one of the following holds:**
+>   * `E` exactly matches `T₁` and `E` does not exactly match `T₂`
+>   * `E` exactly matches both or neither of `T₁` and `T₂`, and `T₁` is a [*better conversion target*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11646-better-conversion-target) than `T₂`
+> * `E` is a method group, ...
+
+We add a new definition for ***better collection conversion from expression***, as follows:
+
+Given:
+- `E` is a collection expression with element expressions `[EL₁, EL₂, ..., ELₙ]`
```
Comment (cston) [inline proposals/collection-expression-better-conversion.md:24]: If `EL` are meant to include spread elements, then we should avoid calling these expressions.
```suggestion
- `E` is a collection expression with elements `[EL₁, EL₂, ..., ELₙ]`
```

File: proposals/collection-expression-better-conversion.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+- `T'₂` is implicitly convertible to `T'₁`, and `T'₁` is not implicitly convertible to `T'₂`.
+  
+Otherwise, `T'₁` neither as good as nor better than `T₂`.
+
+If `E₁` and `E₂` have an identity conversion, then the element conversions are as good as each other. Otherwise, the element conversions to `E₁` are better than the element conversions to `E₂` if:
+- For every `ELᵢ`, `CE₁ᵢ` is at least as good as `CE₂ᵢ`, and
+- There is at least one i where `CE₁ᵢ` is better than `CE₂ᵢ`
+  
+Otherwise, neither set of element conversions is better than the other, and they are also not as good as each other.  
+Conversion comparisons are made using better conversion from expression if `ELᵢ` is not a spread expression. If `ELᵢ` is a spread expression, we use better conversion from the element type of the spread collection to `E₁` or `E₂`, respectively.
```
Comment (cston) [inline proposals/collection-expression-better-conversion.md:49]: Spreads are not considered expressions.
```suggestion
Conversion comparisons are made using better conversion from expression if `ELᵢ` is not a spread element. If `ELᵢ` is a spread element, we use better conversion from the element type of the spread collection to `E₁` or `E₂`, respectively.
```

File: proposals/collection-expression-better-conversion.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+- `E₁` is the element type of `T₁`
+- `E₂` is the element type of `T₂`
+- `CE₁ᵢ` are the series of conversions from `ELᵢ` to `E₁`
+- `CE₂ᵢ` are the series of conversions from `ELᵢ` to `E₂`
+
+`T'₁` is at least as good as `T'₂` if there is an identity conversion between them.  
+`T'₁` is better than `T'₂` if:
+- `T'₁` is ReadOnlySpan<T>, and `T'₂` is not ReadOnlySpan<T>
+- `T'₂` is Span<T>, and `T'₂` is an *array_or_array_interface*
+- `T'₂` is implicitly convertible to `T'₁`, and `T'₁` is not implicitly convertible to `T'₂`.
```
Comment (cston) [inline proposals/collection-expression-better-conversion.md:40]: Should `T'₁` and `T'₂` be reversed here? For example, we should prefer `List<T>` over `IEnumerable<T>`, and `List<T>` is implicitly convertible to `IEnumerable<T>`.

```suggestion
- `T'₁` is implicitly convertible to `T'₂`, and `T'₂` is not implicitly convertible to `T'₁`.
```

File: proposals/collection-expression-better-conversion.md (hunk 3 above)
Comment (cston) [inline proposals/collection-expression-better-conversion.md:39]:
```suggestion
- `T'₁` is Span<T>, and `T'₂` is an *array_or_array_interface*
```

File: proposals/collection-expression-better-conversion.md (hunk 2 above)
Comment (333fred) [reply proposals/collection-expression-better-conversion.md:36]: I'm specifically using the words "at least as good". I'm not certain that having an identity conversion will remain the only way for two collection types to be as good as each other.

File: proposals/collection-expression-better-conversion.md (hunk 1 above)
Comment (333fred) [reply proposals/collection-expression-better-conversion.md:53]: I used the wording I did specifically here.

**Review by cston — APPROVED** [review]

**Comment by 333fred** [comment]
Going to merge this as is so I have something to link to from the notes. Further refines will be sent as PRs to that file.
